src/Npgsql/Internal/Converters/ArrayConverter.cs

```diff
@@ -449,15 +462,17 @@ ValueTask IElementOperations.Read(bool async, PgReader reader, bool isDbNull, ob
     unsafe ValueTask ReadAsync(PgStreamingConverter<TElement> converter, PgReader reader, object collection, int[] indices, CancellationToken cancellationToken)
     {
         if (converter.ReadAsyncAsTask(reader, cancellationToken, out var result) is { } task)
-            return AwaitTask(task, new(this, &SetResult), collection, indices);
+            return PgArrayConverter.AwaitTask(task, new(this, &SetResult), collection, indices);
 
         SetValue(collection, indices, result);
         return new();
 
-        // Using .Result on ValueTask is equivalent to GetAwaiter().GetResult(), this removes TaskAwaiter<TElement> rooting.
         static void SetResult(Task task, object collection, int[] indices)
         {
-            SetValue(collection, indices, new ValueTask<TElement>((Task<TElement>)task).Result);
+            // Justification: exact type Unsafe.As used to reduce generic duplication cost.
+            Debug.Assert(task is Task<TElement>);
+            // Using .Result on ValueTask is equivalent to GetAwaiter().GetResult(), this removes TaskAwaiter<T> rooting.
```

Comment:
Not new in this PR, but I seem to remember that .Result causes AggregateException to be thrown on failure, wrapping the actual exceptionn (whereas GetAwaiter().GetResult() unwraps)... May be worth making sure.

Reply:
As I understand ValueTask doesn't do so due to the expectation that it's only used for sequential async flows (single consumer single producer). As opposed to Task supporting Dataflow, Task.WhenAll and friends, where exceptions from multiple tasks can come together in one.